File: Algo/Leetcode/072EditDistance.cpp
// O(mn) space and time
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n = word1.length();
        int m = word2.length();
        if(n==0 && m==0) return 0;
        if(n==0 || m==0) return n==0?m:n;
        vector<vector<int>> dp(n+1, vector<int>(m+1));
        for(int i=0; i<=n; i++) dp[i][0] = i;
        for(int j=0; j<=m; j++) dp[0][j] = j;
        for(int i=1; i<=n; i++){
            for(int j=1; j<=m; j++){
                dp[i][j] = min(dp[i-1][j-1] + (word1[i-1] != word2[j-1]), min(dp[i-1][j], dp[i][j-1]) + 1);
            }
        }
        return dp[n][m];
    }
};

File: Algo/Leetcode/072EditDistance.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        n = len(word1)
        m = len(word2)

        dp = [[0 for _ in range(m+1)] for _ in range(n+1)]

        for i in range(n+1):
            dp[i][0] = i
        for j in range(m+1):
            dp[0][j] = j

        for i in range(1, n+1):
            for j in range(1, m+1):
                dp[i][j] = min(min(dp[i][j-1]+1, dp[i-1][j]+1), dp[i-1][j-1]+int(word1[i-1]!=word2[j-1]))

        return dp[n][m]